### PR TITLE
[SPARK-32712][SQL] Support writing Hive bucketed table (Hive file formats with Hive hash)

### PR DESCRIPTION
**Pull request by c21**
### What changes were proposed in this pull request?

This is to support writing Hive bucketed table with Hive file formats (the code path for Hive table write - `InsertIntoHiveTable`). The bucketed table is partitioned with Hive hash, same as Hive, Presto and Trino.

### Why are the changes needed?

To make Spark write other-SQL-engines-compatible bucketed table. Same motivation as https://github.com/apache/spark/pull/33432 .

### Does this PR introduce _any_ user-facing change?

Yes. Before this PR, writing to these Hive bucketed table would throw an exception in Spark if config "hive.enforce.bucketing" or "hive.enforce.sorting" set to true. After this PR, writing to these Hive bucketed table would succeed. The table can be read back by Presto and Trino efficiently as other Hive bucketed table.

### How was this patch tested?

Modified unit test in `BucketedWriteWithHiveSupportSuite.scala`, to verify bucket file names and each row in each bucket is written properly, for Hive write code path as well.

### CODE REVIEW

File: sql/hive/src/test/scala/org/apache/spark/sql/sources/BucketedWriteWithHiveSupportSuite.scala [hunk 1]
```diff
@@ -48,29 +49,37 @@ class BucketedWriteWithHiveSupportSuite extends BucketedWriteSuite with TestHive
...
-          testBucketing(
-            new File(tableDir(table), s"k=$k"),
-            format,
-            8,
-            Seq("i", "j"),
-            Seq("i"),
-            df,
-            bucketIdExpression,
-            getBucketIdFromFileName)
+            withSQLConf("hive.exec.dynamic.partition.mode" -> "nonstrict") {
```
Comment (c21) [inline sql/hive/src/test/scala/org/apache/spark/sql/sources/BucketedWriteWithHiveSupportSuite.scala:67]: This is added as Hive write code path enforces it - https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/InsertIntoHiveTable.scala#L161 .

**Comment by c21** [comment]
@cloud-fan - could you help take a look when you have time? Thanks.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by c21** [comment]
Thank you @cloud-fan for review!

**Pull request closed without merging**